### PR TITLE
Fix an assert in canonicalizeFunctionArgument.

### PR DESCRIPTION
**Pull request by atrick**
Assertion failed: (succeed && "should be filtered by FindBorrowScopeUses"), function canonicalizeFunctionArgument, file CanonicalizeBorrowScope.cpp, line 798

Canonicalization for guaranteed function arguments is triggered by
SILCombine without any up-front analysis. Because the canonicalization
rewrites the function argument's copies in place, it must always
succeed.

Fix the visitBorrowScopeUses utility to be aware that it is being
invoked on a function argument and avoid bailing out.

---

To make this problem reproduce reliably, I enabled it in the CopyPropagation pass. This way, the code in question runs over all SIL. Previously, it happened to be invoked during certain SILCombine optimization and could only assert in a rare case involving unmanaged references.

### CODE REVIEW

**Comment by atrick** [comment]
@swift-ci smoke test

File: test/SILOptimizer/copy_propagation_borrow.sil [hunk 1]
```diff
@@ -1178,3 +1179,24 @@ bb0(%0 : @guaranteed $String):
...
+
+// visitBorrowScope first looks through this copy then sees a
+// forwarding operation (ref_to_unmanaged). It generally cannot prove
+// that all uses of an unowned forward are within the borrow
+// scope. This is still safe, however, because the borrowed value is a
+// function argument. In fact, borrow scope rewriting *must* succeed
+// because it begins rewriting guaranteed function arguments without
+// checking the uses first.
+//
+// CHECK-LABEL-NOSCOPES: sil [ossa] @testEscapingGuaranteedToOwned : $@convention(method) (@guaranteed C) -> () {
```
Comment (nate-chandler) [inline test/SILOptimizer/copy_propagation_borrow.sil:1191]: Should this be `CHECK-NOSCOPES-LABEL`?

File: lib/SILOptimizer/Utils/CanonicalizeBorrowScope.cpp [hunk 2]
```diff
@@ -261,8 +263,13 @@ bool CanonicalizeBorrowScope::visitBorrowScopeUses(SILValue innerValue,
         // Pointer escapes are only allowed if they use the guaranteed value,
         // which means that the escaped value must be confied to the current
         // borrow scope.
-        if (use->get().getOwnershipKind() != OwnershipKind::Guaranteed)
+        if (use->get().getOwnershipKind() != OwnershipKind::Guaranteed
+            && !isa<SILFunctionArgument>(innerValue)) {
+          return false;
+        }
+        if (!visitor.visitUse(use)) {
```
Comment (nate-chandler) [inline lib/SILOptimizer/Utils/CanonicalizeBorrowScope.cpp:270]: Would asserting inside the condition that `!isa<SILFunctionArgument>(innerValue)` be worthwhile?

**Review by nate-chandler — APPROVED** [review]
Great to be doing borrow canonicalizations for all guaranteed function arguments!

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
SwiftREPL/BreakpointSimple.test

**Comment by atrick** [comment]
@swift-ci smoke test and merge

**Pull request merged**